src/transaction/log_replication.cpp

```diff
@@ -1,332 +1,498 @@
+	  case LOG_COMPENSATE:
+	    read_and_redo_record<log_rec_compensate> (thread_entry, header.type, m_redo_lsa);
+	    break;
+	  case LOG_DBEXTERN_REDO_DATA:
+	  {
+	    const log_rec_dbout_redo dbout_redo = m_reader.reinterpret_copy_and_add_align<log_rec_dbout_redo> ();
+	    log_rcv rcv;
+	    rcv.length = dbout_redo.length;
+
+	    log_rv_redo_record (&thread_entry, m_reader, RV_fun[dbout_redo.rcvindex].redofun, &rcv, &m_redo_lsa, 0,
+				nullptr, m_redo_unzip);
+	    break;
+	  }
+	  case LOG_COMMIT:
+	    calculate_replication_delay_or_dispatch_async<log_rec_donetime> (
+		    thread_entry, m_redo_lsa);
+	    break;
+	  case LOG_ABORT:
+	    calculate_replication_delay_or_dispatch_async<log_rec_donetime> (
+		    thread_entry, m_redo_lsa);
+	    break;
+	  case LOG_DUMMY_HA_SERVER_STATE:
+	    calculate_replication_delay_or_dispatch_async<log_rec_ha_server_state> (
+		    thread_entry, m_redo_lsa);
+	    break;
+	  default:
+	    // do nothing
+	    break;
+	  }
+
+	{
+	  std::unique_lock<std::mutex> lock (m_redo_lsa_mutex);
+	  m_redo_lsa = header.forw_lsa;
+	}
+	if (m_parallel_replication_redo != nullptr)
+	  {
+	    m_minimum_log_lsa->set_for_outer (m_redo_lsa);
+	  }
+
+	// to accurately track progress and avoid clients to wait for too long, notify each change
+	m_redo_lsa_condvar.notify_all ();
+
+	m_perfmon_redo_sync.track_and_start ();
+      }
+  }
+
+  template <typename T>
+  void
+  replicator::read_and_redo_btree_stats (cubthread::entry &thread_entry, LOG_RECTYPE rectype, const log_lsa &rec_lsa,
+					 const T &log_rec)
+  {
+    //
+    // Recovery redo does not apply b-tree stats directly into the b-tree root page. But while replicating on the page
+    // server, we have to update the statistics directly into the root page, because it may be fetched by a transaction
+    // server and stats have to be up-to-date at all times.
+    //
+    // To redo the change directly into the root page, we need to simulate having a redo job on the page and we need
+    // the page VPID. The VPID is obtained from the redo data of the log record. Therefore, the redo data must be read
+    // first, then a special job is created with all required information.
+    //
+
+    // Get redo data and read it
+    LOG_RCV rcv;
+    rcv.length = log_rv_get_log_rec_redo_length<T> (log_rec);
+    if (log_rv_get_log_rec_redo_data (&thread_entry, m_reader, log_rec, rcv, rectype, m_undo_unzip, m_redo_unzip)
+	!= NO_ERROR)
+      {
+	logpb_fatal_error (&thread_entry, true, ARG_FILE_LINE, "replicator::read_and_redo_btree_stats");
+	return;
+      }
+    BTID btid;
+    log_unique_stats stats;
+    const char *datap = rcv.data;
+    OR_GET_BTID (datap, &btid);
+    datap += OR_BTID_ALIGNED_SIZE;
+    stats.num_keys = OR_GET_INT (datap);
+    datap += OR_INT_SIZE;
+    stats.num_oids = OR_GET_INT (datap);
+    datap += OR_INT_SIZE;
+    stats.num_nulls = OR_GET_INT (datap);
+    datap += OR_INT_SIZE;
+    VPID root_vpid = { btid.root_pageid, btid.vfid.volid };
```

Comment:
A template function should be extracted up to here and moved to `log_recovery_redo.hpp`

Reply:
I see the template functions in log_recovery_redo.hpp as generally applicable to all log records of a given type. The log_rv_get_btree_unique_stats function is applicable only to a specific recovery index and only to replication, not redo.

One possible improvement would be to group the redo data reading part into a function:
```c
void
btree_rv_read_stats_data (const char *data, BTID &btid, log_unique_stats &stats)
{
    BTID btid;
    log_unique_stats stats;
    const char *datap = data;
    OR_GET_BTID (datap, &btid);
    datap += OR_BTID_ALIGNED_SIZE;
    stats.num_keys = OR_GET_INT (datap);
    datap += OR_INT_SIZE;
    stats.num_oids = OR_GET_INT (datap);
    datap += OR_INT_SIZE;
    stats.num_nulls = OR_GET_INT (datap);
    datap += OR_INT_SIZE;
}
```
This chunk of code is repeated in three other functions (btree_rv_update_tran_stats, btree_rv_undo_global_unique_stats_commit and btree_rv_redo_global_unique_stats_commit).

My suggestion is to add btree_rv_read_stats_data and use it everywhere applicabe. read_and_redo_btree_stats becomes:
```c
    // Get redo data and read it
    LOG_RCV rcv;
    rcv.length = log_rv_get_log_rec_redo_length<T> (log_rec);
    if (log_rv_get_log_rec_redo_data (&thread_entry, m_reader, log_rec, rcv, rectype, m_undo_unzip, m_redo_unzip)
	!= NO_ERROR)
      {
	logpb_fatal_error (&thread_entry, true, ARG_FILE_LINE, "replicator::read_and_redo_btree_stats");
	return;
      }
    
    BTID btid;
    log_unique_stats stats;
    btree_rv_read_stats_data (rcv.data, btid, stats);
    VPID root_vpid = { btid.root_pageid, btid.vfid.volid };
    
    // Create a job or apply the change immediately
    if (m_parallel_replication_redo)
      {
	auto job = std::make_unique<redo_job_btree_stats> (root_vpid, rec_lsa, stats);
	m_parallel_replication_redo->add (std::move (job));
      }
    else
      {
	replicate_btree_stats (thread_entry, root_vpid, stats, rec_lsa);
      }
```

What do you think?